### PR TITLE
[SPARK-10779][PYSPARK][MLLIB] Set initialModel for KMeans model in PySpark (spark.mllib)

### PR DESCRIPTION
**Pull request by evanyc15**
Provide initialModel param for pyspark.mllib.clustering.KMeans

### CODE REVIEW

**Comment by jkbradley** [comment]
ok to test

File: python/pyspark/mllib/clustering.py [hunk 1]
```diff
@@ -90,6 +90,14 @@ class KMeansModel(Saveable, Loader):
     ...     rmtree(path)
     ... except OSError:
     ...     pass
+
+    >>> data = array([-383.1,-382.9, 28.7,31.2, 366.2,367.3]).reshape(3, 2)
+    >>> model = KMeans.train(
+    ...     sc.parallelize(data), 3, maxIterations=0, runs=1, initializationMode="random",
+    ...     seed=50, initializationSteps=5, epsilon=1e-4,
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: No need to set many of these parameters for this test; remove unnecessary ones.

File: python/pyspark/mllib/clustering.py [hunk 2]
```diff
@@ -144,10 +152,14 @@ class KMeans(object):
 
     @classmethod
     def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||",
-              seed=None, initializationSteps=5, epsilon=1e-4):
+              seed=None, initializationSteps=5, epsilon=1e-4, initialModel=None):
         """Train a k-means clustering model."""
+        cluster_initialModel = []
+        if initialModel is not None and isinstance(initialModel, KMeansModel):
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: Raise error if not None but not instance of KMeansModel

**Comment by jkbradley** [comment]
This looks fine other than those 2 comments.  Thanks!

**Comment by evanyc15** [comment]
Hey jkbradley,

I have pushed another commit that removes the excess parameters in the Doctest and adds an if statement for checking if initialModel is not None and not isinstance of KMeansModel

Thanks

**Comment by evanyc15** [comment]
Jenkins test this

File: python/pyspark/mllib/clustering.py [hunk 3]
```diff
@@ -90,6 +90,14 @@ class KMeansModel(Saveable, Loader):
     ...     rmtree(path)
     ... except OSError:
     ...     pass
+
+    >>> data = array([-383.1, 31.2, 32.2, 367.3]).reshape(2, 2)
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: I would recommend keeping the original data structure; I could imagine us later adding checks which print warnings when given 2 data points but 3 cluster centers.

File: python/pyspark/mllib/clustering.py [hunk 4]
```diff
@@ -90,6 +90,14 @@ class KMeansModel(Saveable, Loader):
...
+    >>> model = KMeans.train(
+    ...     sc.parallelize(data), 3, maxIterations=0, runs=1, initializationMode="random",
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: By remove unnecessary parameters, I meant to remove: runs, initializationMode, initializationSteps, epsilon.

File: python/pyspark/mllib/clustering.py [hunk 5]
```diff
@@ -144,10 +152,16 @@ class KMeans(object):
 
     @classmethod
     def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||",
-              seed=None, initializationSteps=5, epsilon=1e-4):
+              seed=None, initializationSteps=5, epsilon=1e-4, initialModel=None):
         """Train a k-means clustering model."""
+        clusterInitialModel = []
+        if initialModel is not None and not isinstance(initialModel, KMeansModel):
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: Can you please use nested if statements to reduce code duplication?

**Comment by evanyc15** [comment]
Jenkins test this

File: python/pyspark/mllib/clustering.py [hunk 6]
```diff
@@ -144,10 +150,16 @@ class KMeans(object):
 
     @classmethod
     def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||",
-              seed=None, initializationSteps=5, epsilon=1e-4):
+              seed=None, initializationSteps=5, epsilon=1e-4, initialModel=None):
         """Train a k-means clustering model."""
+        clusterInitialModel = []
+        if initialModel is not None:
+            if not isinstance(initialModel, KMeansModel):
+                raise Exception("initialModel needs to be of type KMeansModel")
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: Please state what the bad type is in the error: ```type(initialModel)```

File: python/pyspark/mllib/clustering.py [hunk 7]
```diff
@@ -90,6 +90,12 @@ class KMeansModel(Saveable, Loader):
     ...     rmtree(path)
     ... except OSError:
     ...     pass
+
+    >>> data = array([-383.1, -382.9, 28.7, 31.2, 366.2, 367.3]).reshape(3, 2)
+    >>> model = KMeans.train(sc.parallelize(data), 3, maxIterations=0,
+    ...     initialModel = KMeansModel([(-1000.0, -1000.0),(5.0, 5.0),(1000.0, 1000.0)]))
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: nit: Python style (spacing): ```initialModel=KMeansModel([(-1000.0,-1000.0), (5.0,5.0), (1000.0,1000.0)])```

**Comment by jkbradley** [comment]
Just 2 tiny comments.

Btw, only committers can get Jenkins to run tests on demand.  (This helps avoid overtaxing Jenkins.)

**Comment by jkbradley** [comment]
Thanks, I'll go ahead and merge this, though the style still isn't quite what I wrote....  I know style is annoying to match, but it helps a lot with keeping the codebase clean and easy to read.

LGTM, merging with master

**Pull request closed without merging**